src/core/thread/network_data_publisher.cpp

```diff
@@ -0,0 +1,1014 @@
+
+    VerifyOrExit((GetState() == kAdded) || (GetState() == kRemoving));
+
+    switch (GetType())
+    {
+    case kTypeAnycast:
+        SuccessOrExit(Get<Service::Manager>().Remove<Service::DnsSrpAnycast>(
+            Service::DnsSrpAnycast::ServiceData(mInfo.GetSequenceNumber())));
+        break;
+
+    case kTypeUnicast:
+        SuccessOrExit(Get<Service::Manager>().Remove<Service::DnsSrpUnicast>(
+            Service::DnsSrpUnicast::ServiceData(mInfo.GetAddress(), mInfo.GetPort())));
+        break;
+
+    case kTypeUincastMeshLocalEid:
+        SuccessOrExit(Get<Service::Manager>().Remove<Service::DnsSrpUnicast>());
+        break;
+    }
+
+    registerWithLeader = true;
+
+    otLogInfoNetData("Publisher: Removed %s from network data", ToString().AsCString());
+
+exit:
+    SetState(aNextState);
+    return registerWithLeader;
+}
+
+void Publisher::DnsSrpServiceEntry::Process(void)
+{
+    // This method checks the entries currently present in Network Data
+    // based on which it then decides whether or not take action
+    // (add/remove or keep monitoring).
+
+    uint8_t numEntries          = 0;
+    uint8_t numPreferredEntries = 0;
+    uint8_t desiredNumEntries   = 0;
+
+    // Do not make any changes if device is not attached, and wait
+    // for role change event.
+    VerifyOrExit(Get<Mle::Mle>().IsAttached());
+
+    VerifyOrExit(GetState() != kNoEntry);
+
+    switch (GetType())
+    {
+    case kTypeAnycast:
+        CountAnycastEntries(numEntries, numPreferredEntries);
+        desiredNumEntries = kDesiredNumAnycast;
+        break;
+
+    case kTypeUnicast:
+    case kTypeUincastMeshLocalEid:
+        CountUnicastEntries(numEntries, numPreferredEntries);
+        desiredNumEntries = kDesiredNumUnicast;
+        break;
+    }
+
+    UpdateState(numEntries, numPreferredEntries, desiredNumEntries);
+
+exit:
+    return;
+}
+
+void Publisher::DnsSrpServiceEntry::CountAnycastEntries(uint8_t &aNumEntries, uint8_t &aNumPreferredEntries) const
+{
+    // Count the number of matching "DNS/SRP Anycast" service entries
+    // in the Network Data (the match requires the entry to use same
+    // "sequence number" value). We prefer the entries associated with
+    // smaller RLCO16.
+
+    Service::DnsSrpAnycast::ServiceData serviceData(mInfo.GetSequenceNumber());
+    const ServiceTlv *                  serviceTlv = nullptr;
+    const ServerTlv *                   serverSubTlv;
+
+    while ((serviceTlv = Get<Leader>().FindNextMatchingService(serviceTlv, Service::kThreadEnterpriseNumber,
+                                                               reinterpret_cast<const uint8_t *>(&serviceData),
+                                                               serviceData.GetLength())) != nullptr)
+    {
+        for (const NetworkDataTlv *start = serviceTlv->GetSubTlvs();
+             (serverSubTlv = NetworkData::FindTlv<ServerTlv>(start, serviceTlv->GetNext())) != nullptr;
+             start = serverSubTlv->GetNext())
```

Comment:
Adding a note  to update the loops in this module to use `TlvIterator` with https://github.com/openthread/openthread/pull/6621 merged.